### ISSUE TITLE
Non-standard Calendar Support

### ISSUE THREAD

**Issue by jhamman** [numpy/numpy#6207]
The current `datetime64` API only supports the standard Gregorian calendar.  In the climate science community (and presumably others), a number of other calendars systems are used, for example:
- [Julian Calendar](https://en.wikipedia.org/wiki/Julian_calendar)
- [No Leap Day Calendar](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#calendar):  Gregorian calendar without leap years, i.e., all years are 365 days long. 
- [All Leap Day Calendar](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#calendar):  Gregorian calendar with every year being a leap year, i.e., all years are 366 days long.

I'd like to propose adding calendar support to the `datetime64` API with the hope that changes here will trickle down into [pandas](https://github.com/pydata/pandas) and [xray](https://github.com/xray/xray).

I'm not sure exactly how I'd like to see the API and would like to hear from those familiar with `datetime64` on how it could be done.  Of course, the default calendar behavior should remain the Gregorian calendar.

xref: https://github.com/xray/xray/pull/523, https://github.com/pydata/pandas/issues/7307
cc: @shoyer @rabernat @ocefpaf

**Comment by ocefpaf** [comment]
:+1: to this.  I have no idea what would take to implement this in `datetime64`, but I would like to add how [netCDF4-python](https://github.com/Unidata/netcdf4-python/blob/master/netcdftime/netcdftime.py) currently implements this for future reference.

**Comment by njsmith** [comment]
My feeling is that the only realistic way to handle this is for numpy to provide the infrastructure to let you define your own date time dtypes with whatever idiosyncratic behavior your application requires, and then get out of the way and let you get on with it. `datetime64` still has major brokenness at handling the regular civil calendar :-(. This won't get fixed overnight, but we are working on it.

**Comment by rabernat** [comment]
:+1: from me too.

@njsmith - one possible advantage of these non-standard calendars is that they are actually much simpler than the "real" calendar. So they could provide a useful set of test cases.

**Comment by pwolfram** [comment]
:+1:

**Comment by akturner** [comment]
:+1: This would be very useful in the climate modeling community. Several IPCC class models need this functionality to analyze output.

**Comment by mhvk** [comment]
Just for information: `astropy` has a proper `Time` class to which one can easily add ones favourite calendars (including proper accounting of leap seconds, etc., in conversions to julian day, etc.; however, no timezones); see http://docs.astropy.org/en/latest/time/index.html

**Comment by CommonClimate** [comment]
Adding my +1 to this issue. The lack of calendar support essentially prevents me from using all the powerful features of pandas, and I have to treat the time axis as a column in my dataframe, which is bulky and cuts me off from powerful tools like pd.resample.   Beyond the calendars mentioned by @jhamman, support for "BP" (before present) calendars (present being either AD 1950 or AD 2000 depending on the convention) would be very useful for the paleoclimate community.

**Comment by charris** [comment]
Doing a proper job for something as complicated as calenders requires specialist knowlege. It's not rocket science, it's harder ;) The best approach here would be for interested parties to get together, design and API and put together an NEP and some code for a PR.

**Comment by charris** [comment]
Note that numpy time itself is perhaps overdesigned and more complicated than one might like. I'm not sure what we want to do about that.

**Comment by njsmith** [comment]
IMO the only realistic way to handle time properly in numpy is (a) numpy fixes its dtype system to make user-defined parametrized dtypes a reasonable thing to write, (b) a variety of third-party packages implement whatever calendars and in-memory representations are useful to their particular communities. There just isn't any one-size-fits-all solution here, and the numpy core team just doesn't have the appropriate expertise to handle all the different cases.

**Comment by njsmith** [comment]
...I guess I said the same thing upthread already. Oh well :-)

**Comment by CommonClimate** [comment]
An alternate viewpoint is that perhaps it is not a numpy issue, but a pandas issue... Still, I tend to agree with you njsmith, that the ability to customize dtypes would solve a lot of issues.

**Comment by jhamman** [comment]
@njsmith - I'm going to ask a question that I'm sure will be nearly impossible to answer. What is the timeframe for (a)?  1 or 10 years?  Are there actual plans in the works? I ask because if this is one of those things that isn't really ever going to happen, those of us interested in seeing calendars in python improved (pandas, xarray, ect.), may seek a different path forward. I know these things take time but any ballpark estimate might be helpful.

**Comment by mhvk** [comment]
Outsiders perspective: `__numpy_ufunc__` has taken about 5 years now from idea to almost implementation. The realistic way forward is to produce a time package oneself, writing it such that eventually one can use `dtype`.  Please do look at astropy's (see link above); it is based on using two `float64` internally for full precision, and is meant to be easy to extend.

**Comment by njsmith** [comment]
@jhamman: 100% of the blocker is finding programmer-hours to work on it - the work is somewhat tricky but doable and we have consensus on the general path forward. But currently numpy has about enough volunteer energy to keep basic things ticking over and gets zero support from companies etc., so larger changes like this are entirely dependent on someone caring enough to show up and help. So I guess your question seems like the wrong way to think about it to me -- if there is energy to seek any path forward, then there is energy to seek this path forward :-). OTOH there is no one working directly on this right now (though @jaimefrio and @madphysicist have expressed some interest in starting, and I've been trying to find ways to get more (any) resources for the project; if we had said resources then this dtype thing is pretty much our #1 priority).

**Comment by njsmith** [comment]
@mhvk: if you're an outsider then everyone's an outsider :-)

**Comment by pv** [comment]
__numpy_ufunc__ took 1 year from idea to implementation --- the delay is 
due to issues not fully anticipated in the initial design. Of course, 
the same may happen with dtypes, but one can be optimistic.

**Comment by shoyer** [comment]
@mhvk I agree that rolling your own time package is the *practical* path forward. But the sad part is that is that without the possibility of creating a numpy dtype, you're stuck rolling all of your own domain specific containers (like astropy) -- no pandas.DataFrame for you.

**Comment by izaid** [comment]
I'm just going to point out that we *do* have user-defined types in DyND, and all of DyND's callables are  overloadable on user-defined types. It took us a lot of time and effort to design a sensible system for that. DyND has other problems for sure, but one practical way forward (depending on what problem people want to solve) could be to help out with DyND.

**Comment by CommonClimate** [comment]
Thanks a lot of all the perspectives. I am shocked to hear from @njsmith that "if we had said resources then this dtype thing is pretty much our #1 priority". What would it take to get you those resources? Google Summer of Code interns, maybe?  Of all the things CS kids spend their summer on, this one would seem worthwhile.  I could also reach out to professional Pythonistas at national labs (LLNL comes to mind). Being relatively new at Python, it's not something that I could do myself, but I am good at finding funding for important projects, and this is very high priority for me.  PM me.

**Comment by shoyer** [comment]
I think this is too hard for a GSOC project. This needs someone who is comfortable getting their hands dirty with the messy C code in the heart of NumPy.

**Comment by CommonClimate** [comment]
I hope such a person exists and has time to work on this!

**Comment by njsmith** [comment]
@CommonClimate:
> Being relatively new at Python, it's not something that I could do myself, but I am good at finding funding for important projects, and this is very high priority for me. PM me.

I don't think github has any PM facility and your profile's pretty anonymous, but if you want to talk offline you can reach me at njs@berkeley.edu
